include/cuco/detail/static_map/kernels.cuh

```diff
@@ -154,40 +155,25 @@ CUCO_KERNEL __launch_bounds__(BlockSize) void insert_or_apply_shmem(
   shared_map_ref.initialize(block);
   block.sync();
 
-  int32_t num_loop = 0;
   while (idx - thread_idx < n) {
     if constexpr (CGSize == 1) {
       // insert-or-apply into the shared map first
+      int32_t inserted = 0;
       if (idx < n) {
         value_type const& insert_pair = *(first + idx);
-        shared_map_ref.insert_or_apply(insert_pair, op);
+        inserted                      = shared_map_ref.insert_or_apply(insert_pair, op);
       }
-      block.sync();
 
-      // for first pass block_cardinality will be < threshold
-      // we can skip block_reduction to find cardinality
-      if (num_loop != 0) {
-        // find if cardinality exceeds threshold
-        int32_t thread_count = 0;
-        if (idx < n) {
-          auto window_idx = thread_idx;
-          while (window_idx < num_windows) {
-            auto const slot = storage[window_idx][0];
-            if (not cuco::detail::bitwise_compare(slot.first, ref.empty_key_sentinel())) {
-              thread_count += 1;
-            }
-            window_idx += BlockSize;
-          }
-        }
-
-        int32_t local_cardinality = BlockReduce(temp_storage).Sum(thread_count);
-        if (thread_idx == 0) block_cardinality = local_cardinality;
-        block.sync();
-
-        if (idx < n) {
-          if (block_cardinality > BlockSize) { break; }
+      if (idx - warp_thread_idx < n)  // all threads in warp particpate
+      {
+        warp.sync();  // sync for inserted to materalize
+        for (int32_t i = warp.size() / 2; i > 0; i /= 2) {
+          inserted += warp.shfl_down(inserted, i);
```

Comment:
could we use `cg::reduce` (https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#reduce) for better performance?

Reply:
We only need the value from the first lane of every CG so we don't even need a full warp reduce.